sw/device/lib/dif/dif_aes.c

```diff
@@ -300,6 +317,87 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
   return kDifOk;
 }
 
+dif_result_t dif_aes_load_aad(const dif_aes_t *aes, const dif_aes_aad_t aad,
+                              bool set_aad_mode) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (set_aad_mode) {
+    // Configure valid AAD bytes.
+    uint32_t reg = bitfield_field32_write(
+        0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, aad.valid_bytes);
+    // Put AES-GCM into AAD mode.
+    reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
+                                 AES_CTRL_GCM_SHADOWED_PHASE_VALUE_GCM_AAD);
+    aes_shadowed_write(aes->base_addr, AES_CTRL_GCM_SHADOWED_REG_OFFSET, reg);
+  }
+
+  if (!aes_input_ready(aes)) {
+    return kDifUnavailable;
+  }
+
+  aes_set_multireg(aes, &aad.data[0], AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_set_gcm_text_mode(const dif_aes_t *aes,
+                                       size_t valid_bytes) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (!aes_idle(aes)) {
+    return kDifUnavailable;
+  }
+
+  // Configure valid TEXT bytes.
+  uint32_t reg = bitfield_field32_write(
+      0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, valid_bytes);
+  // Put AES-GCM into TEXT mode.
+  reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
+                               AES_CTRL_GCM_SHADOWED_PHASE_VALUE_GCM_TEXT);
+  aes_shadowed_write(aes->base_addr, AES_CTRL_GCM_SHADOWED_REG_OFFSET, reg);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_set_gcm_tag_mode(const dif_aes_t *aes, uint64_t len_ptx,
+                                      uint64_t len_aad) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (!aes_idle(aes)) {
+    return kDifUnavailable;
+  }
+
+  // Configure valid TAG bytes.
+  uint32_t reg = bitfield_field32_write(
+      0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, 16);
+  // Put AES-GCM into TAG mode.
+  reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
+                               AES_CTRL_GCM_SHADOWED_PHASE_VALUE_GCM_TAG);
+  aes_shadowed_write(aes->base_addr, AES_CTRL_GCM_SHADOWED_REG_OFFSET, reg);
+
+  // len_aad_ptx = len(aad) || len(ptx)
+  dif_aes_data_t len_aad_ptx;
+  len_aad_ptx.data[0] = __builtin_bswap32((uint32_t)(len_aad >> 32));
+  len_aad_ptx.data[1] = __builtin_bswap32((uint32_t)(len_aad & 0xFFFFFFFF));
+  len_aad_ptx.data[2] = __builtin_bswap32((uint32_t)(len_ptx >> 32));
+  len_aad_ptx.data[3] = __builtin_bswap32((uint32_t)(len_ptx & 0xFFFFFFFF));
+
+  aes_set_multireg(aes, &len_aad_ptx.data[0], AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  // Wait until tag generation has completed.
+  AES_WAIT_FOR_STATUS(aes, AES_STATUS_IDLE_BIT, true);
```

Comment:
I would take all the data handling out. Because as is, this will only work in automatic mode. In manual mode you'll have to set the trigger which this currently does. It's probably better to have a function to configure the phase together with the number of valid bytes. And then something to drive the input data registers.

For tag, you can use a dedicated function to do the byte  swap and write the data registers. But you'll need to be able to send the trigger before reading the tag in manual mode.